### ISSUE TITLE
Can we make the test suite faster?

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#1946]
it seems that the current the test suite is really quite slow.
It seems to take an hour to complete even WITHOUT the --runslow option?
```
============================================== 116 passed, 118 skipped, 66517 warnings in 3798.36s (1:03:18) ==============================================
```

I think this make me and others less keen to run tests at all because it just takes so long.
But this is debatable
How long is "reasonable" for the main test suite to run? 10 minutes?
What's the solution? skipping more?

Crazy idea: is it possible that some recent development that we've done (e.g. autodiscovery of resources) could be slowing down each and every query? Maybe some better caching would speed everything up tremendously?

I have no solutions to offer, but maybe we should discuss at the next AHM and set someone to do some profiling to determine where the bottlenecks are?

**Comment by amykglen** [comment]
I agree it seems like some kind of systemic slowness was introduced in the last year, which certainly could be related to our autodiscovery of KPs, and/or who knows what else. perhaps also DTD is contributing some to slowness? #1932 

10 minutes sounds like a great target for a core test suite.

to get us started here are the 30 slowest 'fast' tests in our pytest suite (run on my laptop):
```
11244.74s call     test_ARAX_json_queries.py::test_drug_disease_query
2691.39s call     test_ARAX_standup_queries.py::test_diseases_treated_by_drug_issue_20
366.90s call     test_ARAX_translate.py::test_overlay_after_lookup
303.18s call     test_ARAX_translate.py::test_lookup
154.62s call     test_ARAX_workflows.py::test_exclude
120.39s call     test_ARAX_translate.py::test_score
108.21s call     test_ARAX_connect.py::test_connect_acetaminophen_to_ptgs1
105.86s call     test_ARAX_standup_queries.py::test_chemical_substances_that_down_regulate_STK11_issue_28
72.34s call     test_ARAX_standup_queries.py::test_chemicals_to_gene_issue_10
70.11s call     test_ARAX_standup_queries.py::test_named_thing_associated_with_acrocynaosis_issue_12
68.35s call     test_ARAX_translate.py::test_fill_success
64.67s call     test_ARAX_expand.py::test_edge_constraints
55.15s call     test_ARAX_standup_queries.py::test_gene_to_pathway_issue_9
51.37s call     test_ARAX_workflows.py::test_FET_ranking_2
45.55s call     test_ARAX_expand.py::test_many_kp_query
44.29s call     test_ARAX_translate.py::test_fill_error
41.88s call     test_ARAX_json_queries.py::test_ngd_added
41.59s call     test_ARAX_overlay.py::test_FET_ex1
41.03s call     test_ARAX_overlay.py::test_jaccard
39.49s call     test_ARAX_filter_kg.py::test_edge_key_removal
38.74s call     test_ARAX_expand.py::test_exclude_edge_no_results
37.17s call     test_ARAX_query.py::test_query_by_query_graph_2
37.07s call     test_ARAX_json_queries.py::test_query_by_query_graph_2
37.00s call     test_ARAX_expand.py::test_option_group_query_one_hop
36.53s call     test_ARAX_expand.py::test_xdtd_different_predicates
36.48s call     test_ARAX_translate.py::test_bind
30.22s call     test_ARAX_resultify.py::test_node_binding_query_id_one_hop_single_input_curie
30.05s call     test_ARAX_translate.py::test_complete_results
29.60s call     test_ARAX_filter_kg.py::test_warnings
29.57s call     test_ARAX_translate.py::test_filter_results_top_n
```
the longest test took 3 hours apparently! and the second longest was 45 minutes. I think those two tests both took a long time on the DTD step. I think even longer than before we rolled out the new DTD database yesterday evening?

**Comment by amykglen** [comment]
of note, it looks like the test suite currently takes too long for our automated testing system to complete (it times out), so our build is showing as failing in github

**Comment by edeutsch** [comment]
another idea: instead of performing 50 from-scratch queries and asserting one or two things in each, might it be possible to perform only a handful of from-scratch queries and then test/assert a lot of things (very quickly) about those very few results?

**Comment by amykglen** [comment]
yeah, great idea, I was thinking the same. think we could drastically reduce our number of tests by doing that.

**Comment by amykglen** [comment]
hey @chunyuma @dkoslicki - just a heads up that I just marked two pytests that involve DTD as 'slow' (see above commit) in an attempt to try to stop `cicd.rtx.ai` from timing out (and thus stop showing our build as failing). apologies for any trouble

**Comment by chunyuma** [comment]
Hey @amykglen, let's see if increasing the `DTD expand` speed in issue #1932 can make these two pytests run faster.

**Comment by chunyuma** [comment]
Hi @amykglen, I've significantly reduced the test time of top 3 "slow" tests that you reported above:
```
3m28.225s call     test_ARAX_json_queries.py::test_drug_disease_query
1m10.596s call     test_ARAX_standup_queries.py::test_diseases_treated_by_drug_issue_20
1m18.035s call     test_ARAX_translate.py::test_overlay_after_lookup
```

Considering they still need more than 1 minute, I would keep marking them as slow tests in pytest. Please let me know anything else I can help for improving the speed.

**Comment by edeutsch** [comment]
So it seems like the category manager may be the cause. When I run a NodeSynonymizer test for get_canonical_curies() I get:
```
==== Get canonical curies for a set of input curies ============================
WARNING: returned with status 404 while retrieving ancestors for biolink:SmallMolecule
WARNING: returned with status 404 while retrieving ancestors for biolink:Disease
WARNING: returned with status 404 while retrieving ancestors for biolink:PhenotypicFeature
WARNING: returned with status 404 while retrieving ancestors for biolink:AnatomicalEntity
WARNING: returned with status 404 while retrieving ancestors for biolink:Protein
WARNING: returned with status 404 while retrieving ancestors for biolink:Gene
WARNING: returned with status 404 while retrieving ancestors for biolink:SmallMolecule
WARNING: returned with status 404 while retrieving ancestors for biolink:Disease
WARNING: returned with status 404 while retrieving ancestors for biolink:PhenotypicFeature
...
Elapsed time: 5.717957079410553
```

So it looks like NodeSynonymizer is slow because it wants category information from the CategoryManager and the CategoryManager is thrashing away at an obsolete endpoint?

Is the CategoryManager still an important thing? or has it been replaced by something?

**Comment by edeutsch** [comment]
So the problem is definitely that the CategoryManager is struggling and slow. One problem is that we internalize our (ARAX's) Biolink version as 3.0.3. And the RENCI Biolink lookup service only supports these versions:
```
$ curl -X 'GET' 'https://bl-lookup-sri.renci.org/versions' -H 'accept: application/json'
["v3.1.2","v3.1.1","v2.4.8","v2.4.7","v2.4.6","v2.4.5","v.2.4.4","v2.4.3","v2.4.2","v2.4.1","v2.4.0","v2.3.1","v2.3.0","v2.2.16","2.2.16","v2.2.15","v2.2.14","2.2.13","2.2.12","2.2.11","2.2.10","2.2.9","2.2.8","2.2.7","2.2.6","2.2.5","2.2.4","2.2.3","2.2.2","2.2.1","2.2.0","2.1.0","2.0.2","2.0.1","2.0.0","1.8.2","1.8.1","1.8.0","1.7.0","1.6.1","1.6.0","1.5.0","1.4.0","1.3.9","latest"]
```

which does not include what we think our Biolink version is. But there are other issues as well.

So I think the CategoryManager is broken.

But I also notice that there seems to be a very nice (and complex) ARAX BiolinkHelper module that I don't know at all.

Can I phone a friend that knows all about the BiolinkHelper to decide what to do with the CategoryManager? I'm thinking that BiolinkHelper might already do everything  that CategoryManager does and we can retire CategoryManager?

In any case, the current state is that NodeSynonymizer is very slow because it calls CategoryManager and CategoryManager is thrashing about unable to function properly. And is probably too slow even if it did function correctly. We need a better solution and I bet someone out there knows how to do it! Help!

**Comment by amykglen** [comment]
yes, I believe `BiolinkHelper` should have all the functionality you need - other modules have switched to using it instead of the `CategoryManager`. we should probably retire it after verifying all usages have been replaced with `BiolinkHelper`.

you use it like this:
```
bh = BiolinkHelper()
bh.get_ancestors("biolink:Drug")
bh.get_descendants("biolink:ChemicalEntity")
```
it automatically detects and uses ARAX's current Biolink version (parsed from the OpenAPI yaml file), though you can also specify a different version if you want (`BiolinkHelper(biolink_version="3.0.0"`). 

by default, `get_ancestors()` and `get_descendants()` include mixins and ARAX-defined conflations (Gene==Protein and Disease==PhenotypicFeature), though you can turn that behavior off with the `include_mixins` and `include_conflations` parameters.

you can also pass in lists of categories/predicates.

there's a readme here with a little more info: https://github.com/RTXteam/RTX/tree/master/code/ARAX/BiolinkHelper

**Comment by edeutsch** [comment]
Hi everyone, an update on this.
Upon first testing the NodeSynonymizer slowness problems identified here:

1. NodeSynonymizer was taking 5 seconds to execute a fairly simple fetch. I determined that the CategoryManager() was trying to fetch category ancestors from an SRI endpoint that wasn't working for us, mostly because they don't support the Biolink version we're using and maybe other problems. This was solved by deprecating the CategoryManager and switching to BiolinkHelper completely. Note that the CategoryManager() was using the BiolinkHelper() already, mostly to know the current Biolink version.

2. After that fix, the NodeSynonymizer went from 5 seconds on that test to 0.5 seconds! Yay! But, then we realized that the initial problem we were seeing was that the NodeSynonymizer was always taking 0.5 seconds no matter how trivial the task. So, I managed to fix a problem, but it wasn't *the* problem. We were still at 0.5 seconds and it should be much faster.

3. After additional profiling, I narrowed the slowness down to a single line:
```
biolink_helper = BiolinkHelper()
```

whoah!
It turns out that the NodeSynonymizer only needed the BiolinkHelper() rarely. But it seemed innocuous to just always set it up. So I have updated get_canonical_curies() to only create the BioinkHelper() object if it really needs it. Which is only needed sometimes. an now the NodeSynonymizer is back to its former self. Instead of taking 0.5 seconds for any and every query, we're now doing:
```
Elapsed time: 0.0010444819927215576
```
Scorching!

So the lessons are:
1. It might be worth seeing if there is a way to speed up instantiating a BiolinkHelper() object. It seems to take 0.5 seconds. Even if all the caller wants to know "What's our current Biolink version?"
2. Any code instantiating the BiolinkHelper() might choose to only instantiate it when absolutely necessary
3. It would be a good programming practice in general to not instantiate complex objects unless really needed. If only I'd taken programming courses, I would surely have learned this, but instead I figured it out the hard way.

I pushed these fixes to master and deployed to test, beta, kg2beta, devED. We have now gone from:
```
2023-01-18T19:50:01.856687 DEBUG: Sending NodeSynonymizer.get_canonical_curies() a list of 1 curies
2023-01-18T19:50:02.381339 DEBUG: Got response back from NodeSynonymizer
```

to:
```
2023-01-28T07:29:43.747069 DEBUG: Sending NodeSynonymizer.get_canonical_curies() a list of 1 curies
2023-01-28T07:29:43.747468 DEBUG: Got response back from NodeSynonymizer
```

Overall the test query has gone from an average of 57 seconds to 40 seconds. Not a fantastic increase, but substantial enough for tonight.

**Comment by mohsenht** [comment]
Hi @edeutsch ,

Pathfinder queries are currently flagged as slow due to their 2–3 minute execution time. I propose implementing DSL-based queries for Pathfinder and use them in test suite. This would allow us to pass parameters that limit traversal depth, enabling us to verify functional correctness without the overhead of generating high-quality production results.
